### PR TITLE
Track input list for atoms in difficulty manager

### PR DESCRIPTION
**Pull request by ajreynol**
This makes our support for `get-difficulty` faster by tracking the set of input formulas that atoms occur in, which is highly important for benchmarks with many assertions and where many lemmas are added at standard effort.

This is work towards making `get-difficulty` scale on a large benchmark from Certora.

### CODE REVIEW

File: src/theory/difficulty_manager.cpp [hunk 1]
```diff
@@ -48,8 +51,7 @@ void DifficultyManager::getDifficultyMap(std::map<Node, Node>& dmap)
   }
 }
 
-void DifficultyManager::notifyLemma(const context::CDHashMap<Node, Node>& rse,
-                                    Node n)
+void DifficultyManager::notifyLemma(Node n)
```
Comment (nafur) [inline src/theory/difficulty_manager.cpp:54]: Make this a `TNode`?

File: src/theory/relevance_manager.cpp [hunk 2]
```diff
@@ -40,7 +43,7 @@ RelevanceManager::RelevanceManager(Env& env, Valuation val)
 {
   if (options().smt.produceDifficulty)
   {
-    d_dman.reset(new DifficultyManager(userContext(), val));
+    d_dman.reset(new DifficultyManager(this, userContext(), val));
```
Comment (nafur) [inline src/theory/relevance_manager.cpp:46]:
```suggestion
    d_dman = std::make_unique<DifficultyManager>(this, userContext(), val);
```

File: src/theory/relevance_manager.cpp [hunk 3]
```diff
@@ -156,7 +183,29 @@ void RelevanceManager::computeRelevance()
...
+    // if we are in full effort check and fail to justify, then we should
+    // give a failure and set success to false, or otherwise calls to
+    // isRelevant cannot be trusted.
+    if (d_inFullEffortCheck)
+    {
+      std::stringstream serr;
+      serr << "RelevanceManager::computeRelevance: WARNING: failed to justify "
+           << input;
+      Trace("rel-manager") << serr.str() << std::endl;
+      Assert(false) << serr.str();
```
Comment (nafur) [inline src/theory/relevance_manager.cpp:203]:
```suggestion
      Assert(false) << "RelevanceManager::computeRelevance: WARNING: failed to justify " << input;
```

File: src/theory/relevance_manager.cpp [hunk 4]
```diff
@@ -389,6 +440,79 @@ bool RelevanceManager::isRelevant(Node lit)
...
+    TNode exp = getExplanationForRelevantInternal(lit);
+    if (!exp.isNull())
+    {
+      return exp;
+    }
+    // if the first time, we get the list of input formulas the atom occurs in
+    if (index==0)
+    {
+      ilist = getInputListFor(lit, false);
+      if (ilist!=nullptr)
```
Comment (nafur) [inline src/theory/relevance_manager.cpp:466]: format.

File: src/theory/relevance_manager.cpp [hunk 5]
```diff
@@ -389,6 +440,79 @@ bool RelevanceManager::isRelevant(Node lit)
...
+{
+  NodeListMap::const_iterator it = d_atomMap.find(atom);
+  if (it == d_atomMap.end())
+  {
+    if (!doMake)
+    {
+      return nullptr;
+    }
+    d_atomMap[atom] = std::make_shared<NodeList>(userContext());
+    it = d_atomMap.find(atom);
```
Comment (nafur) [inline src/theory/relevance_manager.cpp:511]:
```suggestion
    it = d_atomMap.emplace(atom, std::make_shared<NodeList>(userContext()));
```

File: src/theory/relevance_manager.h [hunk 6]
```diff
@@ -81,6 +81,7 @@ class RelevanceManager : protected EnvObj
   using RlvPairHashFunction = PairHashFunction<Node, uint32_t, std::hash<Node>>;
   using NodeList = context::CDList<Node>;
   using NodeMap = context::CDHashMap<Node, Node>;
+  using NodeListMap = context::CDHashMap<Node, std::shared_ptr<NodeList>>;
```
Comment (nafur) [inline src/theory/relevance_manager.h:84]: Why are we using a `std::shared_ptr` here?

File: src/theory/relevance_manager.cpp (hunk 3 above)
Comment (ajreynol) [reply src/theory/relevance_manager.cpp:203]: This changes the intended behavior, I want a debug assertion failure or a Trace warning in production.

File: src/theory/relevance_manager.h (hunk 6 above)
Comment (ajreynol) [reply src/theory/relevance_manager.h:84]: I'll change to unique.

Comment (ajreynol) [reply src/theory/relevance_manager.h:84]: Actually, this is currently a limitation of our context-dependent data structures, we can't store unique_ptr due to the requirements on the templated classes.

File: src/theory/relevance_manager.cpp (hunk 3 above)
Comment (nafur) [reply src/theory/relevance_manager.cpp:203]: My concern is that this induces an overhead in a (non-tracing) production build.
How about this?
```
      Trace("rel-manager") << "RelevanceManager::computeRelevance: failed to justify " << input << std::endl
      Assert(false) << "RelevanceManager::computeRelevance: failed to justify " << input;
```

File: src/theory/relevance_manager.h (hunk 6 above)
Comment (nafur) [reply src/theory/relevance_manager.h:84]: Ah I see...

Comment (nafur) [inline src/theory/relevance_manager.h:84]: Does it actually make sense to have a `CDList` inside a `CDHashMap`?

**Review by nafur — APPROVED** [review]
Just one question (possibly answered with a simple "yes"), otherwise LGTM.

File: src/theory/relevance_manager.cpp (hunk 3 above)
Comment (ajreynol) [reply src/theory/relevance_manager.cpp:203]: This code really should never be run, so performance is not an issue here.

**Pull request merged**